nvtabular/loader/backend.py

```diff
@@ -349,72 +312,80 @@ def create_tensors(self, gdf, cat_names=None, cont_names=None, label_names=None)
         raise NotImplementedError()
 
 
-class TensorBatchDatasetItrFactory:
-    """
-        This is the Factory that creates the correct iterator,
-        given a target library.
-
-        Parameters:
-        dataset: NVTabular dataset
-        target: str, target library you want the iterator supported - torch
-        shuffle: bool, shuffle dataset
-        device: int, represents targeted GPU id
-        total_devs: [int], list represents all avialable GPU IDs
-    """
-
-    def create(self, dataset, target, shuffle=False, device=None, total_devs=None, **kwargs):
-        if target in "torch":
-            return TorchTensorBatchDatasetItr(
-                dataset, shuffle=shuffle, device=device, total_devs=total_devs, **kwargs
-            )
-        else:
-            raise ValueError(target)
-
+def _validate_workflows(workflows, cat_names, cont_names, label_names):
+    assert all([isinstance(w, BaseWorkflow) for w in workflows])
+    # TODO: commenting out until it's clearer what the
+    # columns in workflow.columns_cts["final"]["ctx"] mean
+    # for workflow in workflows:
+    #     assert set(workflow.columns_ctx["categorical"]["base"]) == set(cat_names)
+    #     assert set(workflow.columns_ctx["continuous"]["base"]) == set(cont_names)
+    #     assert set(workflow.columns_ctx["label"]["base"]) == set(label_names)
 
-class TorchTensorBatchDatasetItr(TensorBatchDatasetItr):
-    """
-        For PyTorch Only:
-        Takes input of an NVTabular dataset
-        and creates TorchTensorBatchDatasetItr
-        supplying user defined size chunks.
+    #     cat_names = workflow.columns_ctx["final"]["ctx"]["categorical"]
+    #     cont_names = workflow.columns_ctx["final"]["ctx"]["continuous"]
+    #     label_name = workflow.columns_ctx["final"]["ctx"]["label"][0]
+    return workflows
 
-    """
-
-    def device_ctx(self, dev):
-        return torch.cuda.device(dev)
 
-    def gather_indices(self):
-        per_worker = int(len(self.indices) // len(self.total_devs)) + 1
-        worker_id = self.total_devs.index(self.device)
-        start = worker_id * per_worker
-        return self.indices[start : start + per_worker]
-
-    def _to_tensor(self, gdf, dtype=None):
-        if gdf.empty:
-            return
-        dl_pack = self.to_dlpack(gdf)
-        tens = from_dlpack(dl_pack).type(dtype)
-        return tens
-
-    def create_tensors(self, gdf, cat_names=None, cont_names=None, label_names=None):
-        gdf_cats, gdf_conts, gdf_label = (
-            gdf[_get_embedding_order(cat_names)],
-            gdf[cont_names],
-            gdf[label_names],
+class DataLoader:
+    def __init__(
+            self,
+            dataset,
+            cat_names,
+            cont_names,
+            label_names,
+            batch_size,
+            shuffle,
+            workflows=None,
+            devices=None
+    ):
+        indices = cp.arange(dataset.to_ddf().npartitions)
```

Comment:
I think the shuffle should be done here. That way all the iterators will have the same order of indexes randomized or not.

Reply:
Yeah so this is part of what I was hoping to understand about the index shuffle: does it need to be done at the epoch level, or can we just do it up front? Let's talk later today about what the partition index shuffle is doing, I'll admit my Dask understanding is pretty limited

Reply:
So actually doing it at the per epoch level is better. Introduces more randomness. So to that point we probably should not do it here.